### PR TITLE
Plotting / support complex numbers in plot_by_id

### PR DESCRIPTION
**Pull request by WilliamHPNielsen**
Support the plotting of complex numbers in `plot_by_id`. Still WIP.

Changes proposed in this pull request:
- Add a preparser that splits complex-valued parameters into two real-valued ones

Still pending:

- [x] Actually use the preparser in `plot_by_id`
- [x] Update the documentation to show how to plot complex numbers
- [x] Change `phase_and_mag` into `mag_and_phase`

@QCoDeS/core

### CODE REVIEW

File: qcodes/dataset/plotting.py [hunk 1]
```diff
@@ -269,6 +272,110 @@ def plot_by_id(run_id: int,
...
+        conversion: the conversion method, either "real_and_imag" or
+            "phase_and_mag"
+        degress: whether to return the phase in degrees. The default is to
+            return the phase in radians
+    """
+
+    if conversion not in ['real_and_imag', 'phase_and_mag']:
+        raise ValueError(f'Invalid conversion given. Received {conversion}, '
+                         'but can only accept "real_and_imag" or '
+                         '"phase_and_mag".')
```
Comment (astafan8) [inline qcodes/dataset/plotting.py]: awesome!

`mag_and_phase`? :)

Comment (WilliamHPNielsen) [reply qcodes/dataset/plotting.py]: Hmm, that's probably true. `(r, theta)` sounds natural.

**Comment by WilliamHPNielsen** [comment]
@ThorvaldLarsen, you were one of the first people to request this feature. I'd be good to know if you think that this PR does the job, or whether we overlooked something.

File: qcodes/dataset/plotting.py [hunk 2]
```diff
@@ -269,6 +288,110 @@ def plot_by_id(run_id: int,
...
+    # we build a new NamedData object from the given `alldata` input.
+    # Note that the length of `newdata` will be larger than that of `alldata`
+    # in the case of complex top-level parameters. This is the reason why we
+    # use two variables below,
+
+    for group in alldata:
+        new_group = []
+        new_groups: NamedData = [[], []]
+        for index, parameter in enumerate(group):
+            data: np.ndarray = parameter['data']
```
Comment (astafan8) [inline qcodes/dataset/plotting.py]: `data` seems to be unused, right?
```suggestion
```

File: qcodes/dataset/plotting.py [hunk 3]
```diff
@@ -269,6 +288,110 @@ def plot_by_id(run_id: int,
...
+    new_parameters: List[Dict[str, Union[str, np.ndarray]]] = []
+
+    for n in range(2):
+        d: Dict[str, Union[str, np.ndarray]]
+        d = {'name': new_names[n], 'label': new_labels[n],
+             'unit': new_units[n], 'data': new_data[n]}
+        new_parameters.append(d)
+
+    return tuple(new_parameters)  # type: ignore
+
```
Comment (astafan8) [inline qcodes/dataset/plotting.py]: may i suggest a subjectively simpler implementation of this logic?:
```
new_parameters = (
    {'name': name, 'label': label,
     'unit': unit, 'data': data}
    for name, label, unit, data in zip(
        new_names, new_labels, new_units, new_data)
)

return tuple(new_parameters)
```

File: qcodes/dataset/plotting.py [hunk 4]
```diff
@@ -269,6 +288,110 @@ def plot_by_id(run_id: int,
...
+    if conversion not in ['real_and_imag', 'mag_and_phase']:
+        raise ValueError(f'Invalid conversion given. Received {conversion}, '
+                         'but can only accept "real_and_imag" or '
+                         '"mag_and_phase".')
+
+    newdata = []
+
+    # we build a new NamedData object from the given `alldata` input.
+    # Note that the length of `newdata` will be larger than that of `alldata`
+    # in the case of complex top-level parameters. This is the reason why we
```
Comment (astafan8) [inline qcodes/dataset/plotting.py]: Could you add a clarification phrase after `... in the case of complex top-level parameters.` along the lines of
```
Note that the length of `newdata` will be larger than that of `alldata` 
in the case of complex top-level parameters due to the fact that in this 
function a SINGLE complex-valued top-level parameter gets split into 
TWO real-valued top-level parameters.
```
?

This allows to understand the need for both `new_group` and `new_groups` before reaching the end of the function.

File: qcodes/tests/dataset/test_plotting.py [hunk 5]
```diff
@@ -148,3 +149,97 @@ def test_appropriate_kwargs():
     with _appropriate_kwargs('2D_point', False, **{}) as ap_kwargs:
         assert len(ap_kwargs) == 1
         assert ap_kwargs['cmap'] == qc.config.plotting.default_color_map
+
+
+def test__complex_to_real_preparser():
+
+    # FIRST CASE: one real setpoint, one complex value
```
Comment (astafan8) [inline qcodes/tests/dataset/test_plotting.py]: Shall the `test__complex_to_real_preparser` be split into at least two independent ones (ideally into 4 parts actually)? I'd also suggest for this test to go into its own file.

**Comment by ThorvaldLarsen** [comment]
@WilliamHPNielsen this looks great to me! Thanks for getting it done.

File: qcodes/dataset/plotting.py (hunk 2 above)
Comment (WilliamHPNielsen) [reply qcodes/dataset/plotting.py]: I think we use it to check whether the data values are complex (next line).

File: qcodes/dataset/plotting.py (hunk 3 above)
Comment (WilliamHPNielsen) [reply qcodes/dataset/plotting.py]: I was hoping that all the extra type annotations would somehow be helpful. I guess they're not?

File: qcodes/dataset/plotting.py (hunk 4 above)
Comment (WilliamHPNielsen) [reply qcodes/dataset/plotting.py]: More explanation is always better.

File: qcodes/dataset/plotting.py (hunk 3 above)
Comment (astafan8) [reply qcodes/dataset/plotting.py]: they are but the `for n in range(2):` part kind of kills it, at least for me. Hence, the suggestion.

File: qcodes/dataset/plotting.py (hunk 2 above)
Comment (astafan8) [reply qcodes/dataset/plotting.py]: sono idioto :)

File: qcodes/tests/dataset/test_plotting.py (hunk 5 above)
Comment (WilliamHPNielsen) [reply qcodes/tests/dataset/test_plotting.py]: I agree with splitting it into two tests. I don't think it should go into a separate file unless the preparser also leaves the plotting module (into `plotting_helpers.py` or smth)

File: qcodes/dataset/plotting.py [hunk 6]
```diff
@@ -269,6 +288,120 @@ def plot_by_id(run_id: int,
...
+            new_names, new_labels, new_units, new_data))
+
+    # The reason we ignore the type in the return is that I cannot figure
+    # out how to get mypy to correctly infer the type of iterated values
+    # (the name, label, unit, and data above)
+
+    return new_parameters  # type: ignore
+
+
+def _make_named_data_dict(
```
Comment (astafan8) [inline qcodes/dataset/plotting.py]: unused?

**Comment by jenshnielsen** [comment]
The notebook should use 

load_or_create_experiment and explicitly pass the experiment to the measurement
plot_dataset()

File: qcodes/dataset/plotting.py (hunk 6 above)
Comment (astafan8) [reply qcodes/dataset/plotting.py]: removed.

**Comment by astafan8** [comment]
> The notebook should use load_or_create_experiment and explicitly pass the experiment to the measurement

Done. plot_dataset() change I would defer to another PR. so that we change plot_by_id for plot_dataset in the whole qcodes at once.

**Pull request merged**